modules/mediacodec_dec/mediacodec_dec.c

```diff
@@ -0,0 +1,891 @@
+
+    AMediaCodec *codec;
+    AMediaFormat *format;
+
+#endif
+
+    u32 width, height, stride, out_size;
+    u32 pixel_ar, pix_fmt;
+
+    const char *mime;
+
+    u8 chroma_format, luma_bit_depth, chroma_bit_depth;
+    GF_ESD *esd;
+
+    u32 counter;
+    u16 frame_rate;
+
+    Bool inputEOS, outputEOS;
+        
+    //NAL-based specific
+    char *sps, *pps;
+    u32 sps_size, pps_size;
+    u32 nalu_size_length;
+
+    //hevc
+    char *vps;
+    u32 nb_layers, luma_bpp, chroma_bpp, dec_frames;
+    u8 chroma_format_idc;
+    u16 ES_ID;
+    u32 vps_size;
+
+} MCDec;
+
+
+#ifdef GPAC_ANDROID
+u8 sdkInt() 
+{
+    char sdk_str[3] = "0";
+    __system_property_get("ro.build.version.sdk", sdk_str, "0");
+    return atoi(sdk_str);
+}
+#endif
+
+
+u32 androidPixelFormat(u32 pix_fmt) {
+
+    switch(pix_fmt) {
+        case GF_PIXEL_YV12:
+            return 0x32315659;
+
+        case GF_PIXEL_YUV422:
+                return 0x27;
+
+        case GF_PIXEL_YUV444:
+                return 0x28;
+
+        case GF_PIXEL_NV21:
+                return 0x11;
+
+        case GF_PIXEL_RGB_565:
+                return 0x04;
+
+        case GF_PIXEL_YUY2:
+                return 0x14;
+
+        case GF_PIXEL_RGB_32:
+                return 0x2a;
+
+        case GF_PIXEL_RGB_24:
+                return 0x29;
+
+        case GF_PIXEL_YUV444_10:
+        case GF_PIXEL_YV12_10:
+        case GF_PIXEL_YUV422_10:
+        default: 
+            return -1;
+    }
+}
+
+
+//prepend the inBuffer with the start code i.e. 0x00 0x00 0x00 0x01
+void prependStartCode(char *inBuffer, char *outBuffer, u32 *size) 
+{
+    *size += 4;
+    u32 i;
+                              
+    for(i = 0; i < *size; i++) {
+        if(i < 3)
+            outBuffer[i] = 0x00;   
+        
+        else if (i == 3) 
+            outBuffer[i] = 0x01;
+            
+        else  outBuffer[i] = inBuffer[i-4];
+    }
+}
+
+
+void initMediaFormat(MCDec *ctx, AMediaFormat *format) 
+{
+    AMediaFormat_setString(ctx->format, AMEDIAFORMAT_KEY_MIME, ctx->mime);
+    AMediaFormat_setInt32(ctx->format, AMEDIAFORMAT_KEY_WIDTH, ctx->width);
+    AMediaFormat_setInt32(ctx->format, AMEDIAFORMAT_KEY_HEIGHT, ctx->height);
+    AMediaFormat_setInt32(ctx->format, AMEDIAFORMAT_KEY_STRIDE, ctx->stride);  
+    AMediaFormat_setInt32(ctx->format, AMEDIAFORMAT_KEY_COLOR_FORMAT, androidPixelFormat(ctx->pix_fmt));  
+    AMediaFormat_setInt32(ctx->format, AMEDIAFORMAT_KEY_MAX_INPUT_SIZE, ctx->width * ctx->height);
+}
+
+
+
+//todo initialise other items
+static GF_Err MCDec_InitDecoder(MCDec *ctx) {
+
+
+    char *dsi_data = NULL;
+    u32 dsi_data_size = 0;
+
+    //for hevc
+    u32 i, j;
+    GF_HEVCConfig *cfg = NULL;
+
+    ctx->format = AMediaFormat_new();
+
+    if(!ctx->format) {
+        LOGE("AMediaFormat_new failed");
+        return GF_CODEC_NOT_FOUND;
+    }
+
+
+#ifdef GPAC_ANDROID
+
+        ctx->pix_fmt = GF_PIXEL_NV21;
+
+switch (ctx->esd->decoderConfig->objectTypeIndication) {
```

Comment:
Shouldn't we split these cases into functions? MCDec_InitHevcDecoder, ...InitAvcDecoder, ...